### PR TITLE
Developer docs markdown layout

### PR DESCRIPTION
**Pull request by corwintines**
## Description

- Adds docs markdown layout and most components

## Related Issue
https://www.notion.so/efdn/Migrate-templates-to-NextJS-Layouts-5d0d3681c10640d683f75d35a0f221c0

### CODE REVIEW

File: next.config.js [hunk 1]
```diff
@@ -12,6 +12,15 @@ module.exports = (phase, { defaultConfig }) => {
       // supported locales defined in `i18n.config.json`
       locales: i18nConfig.map((lang) => lang.code).sort(),
     },
+    webpack: (config, { isServer }) => {
```
Comment (pettinarip) [inline next.config.js]:
```suggestion
    webpack: (config) => {
```

File: src/lib/utils/toc.ts [hunk 2]
```diff
@@ -91,7 +91,7 @@ const parseSourceToToCItem = ({ label, id }: SourceHeadingItem): ToCItem => ({
  */
 const addHeadingsAsItems = (
   headings: Array<SourceHeadingItem>,
-  h = 2
+  h = 3
```
Comment (pettinarip) [inline src/lib/utils/toc.ts]: Is there a reason why we need to do this change? `2` is the default for most static pages. You can see [the web3 page](https://deploy-preview-114--ethereum-org-fork.netlify.app/web3) for example that is not showing the `h2`s with this change.

I'd say that if we need to do this for the docs pages, we should consume this function overriding the default, like `addHeadingsAsItems(headings, 3)`.

File: src/lib/interfaces.ts [hunk 3]
```diff
@@ -44,6 +44,23 @@ export interface RawQuizzes {
   [key: string]: RawQuiz
 }
 
+/**
+ * Table of Contents
+ */
+export interface ToCItem {
```
Comment (pettinarip) [inline src/lib/interfaces.ts]: Isn't this the same type defined in `types.ts`? We should consume that one.

File: src/components/SideNav.tsx [hunk 4]
```diff
@@ -3,13 +3,12 @@ import { Box, HStack, Icon } from "@chakra-ui/react"
 
 import { motion } from "framer-motion"
 import { MdExpandMore } from "react-icons/md"
-import { useTranslation } from "gatsby-plugin-react-i18next"
 
-import Link, { BaseLink, IProps as ILinkProps } from "./Link"
+import { BaseLink } from "./Link"
```
Comment (pettinarip) [inline src/components/SideNav.tsx]: We should import the `LinkProps` here and use it in the `SideNavLink` component where we haven't defined the prop types yet.

File: src/components/SideNav.tsx [hunk 5]
```diff
@@ -48,7 +47,7 @@ const LinkContainer: React.FC<{ children: ReactNode }> = ({ children }) => {
   )
 }
 
-const SideNavLink: React.FC<ILinkProps> = ({ children, ...props }) => {
+const SideNavLink = ({ children, ...props }) => {
```
Comment (pettinarip) [inline src/components/SideNav.tsx]: Here we should use the `LinkProps`

File: src/components/SectionNav.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import React from "react"
+import { Flex } from "@chakra-ui/react"
+import Emoji from "@/components/Emoji"
+import Translation from "@/components/Translation"
+import OldHeading from "@/components/OldHeading"
+
+export interface IProps {
```
Comment (pettinarip) [inline src/components/SectionNav.tsx]:
```suggestion
export interface SectionNavProps {
```

File: src/layouts/Docs.tsx [hunk 7]
```diff
@@ -1,3 +1,283 @@
...
+import YouTube from "@/components/YouTube"
+
+import {
+  Heading1 as MdHeading1,
+  Heading2 as MdHeading2,
+  Heading3 as MdHeading3,
+  Heading4 as MdHeading4,
+  Paragraph,
+} from "@/components/MdComponents"
+import { useRouter } from "next/router"
```
Comment (pettinarip) [inline src/layouts/Docs.tsx]: import order

File: src/layouts/Docs.tsx [hunk 8]
```diff
@@ -1,3 +1,283 @@
...
+  Divider,
+  Emoji,
+  GlossaryTooltip,
+  InfoBanner,
+  RollupProductDevDoc,
+  SectionNav,
+  YouTube,
+}
+
+export const DocsLayout = ({ children, frontmatter, slug, tocItems }) => {
```
Comment (pettinarip) [inline src/layouts/Docs.tsx]: We are missing props types in this function

File: src/lib/utils/languages.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+export type Languages = {
+    [lang in Lang]: Language
+  }
+  
+export const defaultLanguage: Lang = "en"
+
+// same data as in the `config.json` but indexed by language code
+const languages: Languages = i18nConfig.reduce((result, config) => {
+return { ...result, [config.code]: config }
```
Comment (pettinarip) [inline src/lib/utils/languages.ts]: Seems that prettier didn't run in this file.

**Review by pettinarip — CHANGES_REQUESTED** [review]
@corwintines nice start 💪🏼 

Beside the comments I left, this is how I see most of the docs pages
image
There seem to be some problem with the text color.

File: src/lib/interfaces.ts (hunk 3 above)
Comment (wackerow) [reply src/lib/interfaces.ts]: This may be a remnant of trying to migrate this over to `types` as it was previously in `interfaces`... Agree though, per our latest guidelines we should switch to the `types.ts` instance and remove this

File: src/components/SectionNav.tsx (hunk 6 above)
Comment (wackerow) [reply src/components/SectionNav.tsx]:
```suggestion
export type SectionNavProps = {
```

Per latest TS guideline discussion, no?

File: src/lib/utils/languages.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import type { Direction, Lang } from "@/lib/types"
+
+export interface Language {
+    code: Lang
+    crowdinCode: string
+    name: string
+    localName: string
+    langDir: Direction
+    dateFormat: string
+  }
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Renamed to `I18nLocale` located in `@/lib/types`

File: src/lib/utils/languages.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+export type Languages = {
+    [lang in Lang]: Language
+  }
+  
+export const defaultLanguage: Lang = "en"
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Renamed to `DEFAULT_LOCALE` located in `@/lib/constants`

File: src/lib/utils/languages.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+// same data as in the `config.json` but indexed by language code
+const languages: Languages = i18nConfig.reduce((result, config) => {
+return { ...result, [config.code]: config }
+}, {} as Languages)
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Replaced with `LOCALES_CODES` located in `@/lib/constants`

File: src/lib/utils/languages.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+const buildLangs = (process.env.GATSBY_BUILD_LANGS || "")
+.split(",")
+.filter(Boolean)
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Renamed to `BUILD_LOCALES` in #85, located in `@/lib/constants`

Comment (wackerow) [reply src/lib/utils/languages.ts]: ...and updated to not use `GATSBY`

File: src/lib/utils/languages.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+} = Object.fromEntries(
+  Object.entries(languages).filter(([lang]) => {
+    // BUILD_LANGS === empty means to build all the langs
+    if (!buildLangs.length) {
+    return true
+    }
+
+    return buildLangs.includes(lang)
+  })
+)
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Do we need two different versions of the language list to distinguish between "build" languages and "all supported languages"? I'd think for any build we'd only care about what is actually being included in that build.

I _think_ we can get rid of this in lieu of `LOCALES_CODES` (which as I use this more, I think we should just rename to `LOCALES`)

File: src/lib/utils/languages.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+export const supportedLanguages = Object.keys(languageMetadata) as Array<Lang>
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Same as above... do we need to differentiate between "supported" and "build"?

File: src/lib/utils/languages.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+export const ignoreLanguages = (Object.keys(languages) as Array<Lang>).filter(
+(lang) => !supportedLanguages.includes(lang)
+)
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Same as above... I see this is used in the Gatsby config, I think we can remove this.

File: src/lib/utils/languages.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    crowdinCode: string
+    name: string
+    localName: string
+    langDir: Direction
+    dateFormat: string
+  }
+
+export type Languages = {
+    [lang in Lang]: Language
+  }
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Would suggest we move this to `@/lib/types` under `I18nLocale` and rename to `Locales`... but, do we need this anymore?

File: src/lib/utils/languages.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+export const supportedLanguages = Object.keys(languageMetadata) as Array<Lang>
+
+export const ignoreLanguages = (Object.keys(languages) as Array<Lang>).filter(
+(lang) => !supportedLanguages.includes(lang)
+)
+
+export const isLang = (lang: string): lang is Lang =>
+supportedLanguages.includes(lang as Lang)
```
Comment (wackerow) [inline src/lib/utils/languages.ts]: Per above comments, I think this would be the only helper function left to import... I would suggest confirming we still need it, and moving this to `@/lib/utils/translations`. I think in the end we don't need this file.

File: src/lib/utils/md.ts [hunk 19]
```diff
@@ -38,6 +38,101 @@ const getPostSlugs = (dir: string, files: string[] = []) => {
     "/roadmap/beacon-chain",
     "/roadmap/merge",
     // Developer docs (0/95)
```
Comment (wackerow) [inline src/lib/utils/md.ts:42]:
```suggestion
    // Developer docs (95/95) ✅
```
=)

More practically though, do we need this list at all anymore?? Would suggest it as a task separate to this PR, but I believe once this is in that's **all** the markdown paths, and we can remove the `temporalAllowedPages` logic

File: src/lib/utils/toc.ts (hunk 2 above)
Comment (wackerow) [reply src/lib/utils/toc.ts]: Agree, changing this will remove all the h2's from all ToCs

File: src/lib/interfaces.ts (hunk 3 above)
Comment (pettinarip) [reply src/lib/interfaces.ts]: Ah ok. Yes, we should try to have just one definition.

File: src/components/SectionNav.tsx (hunk 6 above)
Comment (pettinarip) [reply src/components/SectionNav.tsx]: Good call, correct.

File: src/lib/utils/languages.ts (hunk 14 above)
Comment (pettinarip) [reply src/lib/utils/languages.ts]: Good point. That makes sense.

IIRC, in the prev implementation, the reason was that `buildLangs` could be an empty string and in that particular case we loaded all the langs. In that case, the two variables would end up being different. But with the new implementation, both vars are always going to be the same, so we shouldn't have any troubles.

File: src/lib/utils/languages.ts (hunk 16 above)
Comment (pettinarip) [reply src/lib/utils/languages.ts]: Correct. This is handled different now in Next.

File: src/lib/utils/md.ts (hunk 19 above)
Comment (pettinarip) [reply src/lib/utils/md.ts:42]: True. A `disallowlist` would make more sense now.

File: src/lib/utils/toc.ts (hunk 2 above)
Comment (corwintines) [reply src/lib/utils/toc.ts]: Good catch, thanks for pointing out. Will address.

File: next.config.js [hunk 20]
```diff
@@ -20,6 +20,15 @@ module.exports = (phase, { defaultConfig }) => {
   let nextConfig = {
     ...defaultConfig,
     reactStrictMode: true,
+    webpack: (config) => {
+      // For all file types, you should have existing rules here
```
Comment (pettinarip) [inline next.config.js]: What does this means? that we should have a rule for `.json` files as well?

File: public/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashimoto/index.md [hunk 21]
```diff
@@ -275,7 +280,9 @@ _Proof_. Since `P` is a safe prime, then by [Lagrange's Theorem][lagrange] we ha
 
 The order of `x` cannot be `1`, since by Fermat's Little Theorem we have:
 
-<pre>x<sup>P-1</sup> mod P ≡ 1</pre>
+```
+x<sup>P-1</sup> mod P ≡ 1
```
Comment (pettinarip) [inline public/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashimoto/index.md]: Not sure how to properly display this not using tags but we should avoid using tags in this format.
image

File: public/content/translations/de/developers/docs/nodes-and-clients/index.md [hunk 22]
```diff
@@ -288,10 +288,6 @@ Der Sync-Modus, den du auswählst, wird einen Einfluss auf den benötigten Speic
 
 - Hinweis: Erigon kann keinen Fast-Sync durchführen, aber Full-Pruning ist möglich (~500 GB)
 
-![Ein Diagramm, das aufzeigt, dass die Gigabytes für eine Full-Synchronisation ansteigen](./full-sync.png)
```
Comment (pettinarip) [inline public/content/translations/de/developers/docs/nodes-and-clients/index.md]: I guess these images don't exist anymore, right?

File: src/components/SideNav.tsx [hunk 23]
```diff
@@ -133,15 +135,15 @@ const NavLink: React.FC<IPropsNavLink> = ({ item, path }) => {
...
+export interface SideNavProps {
   path: string
 }
 
 // TODO set tree state based on if current path is a child
 // of the given parent. Currently all `path` items default to open
 // and they only collapse when clicked on.
 // e.g. solution: https://github.com/hasura/gatsby-gitbook-starter/blob/5c165af40e48fc55eb06b45b95c84eb64b17ed32/src/components/sidebar/tree.js
-const SideNav: React.FC<IProps> = ({ path }) => {
+const SideNav: React.FC<SideNavProps> = ({ path }) => {
```
Comment (pettinarip) [inline src/components/SideNav.tsx]: Since we are deprecating `React.FC`, now we should do
```suggestion
const SideNav = ({ path }: SideNavProps) => {
```

File: src/components/SideNavMobile.tsx [hunk 24]
```diff
@@ -62,7 +62,7 @@ const LinkContainer: React.FC<{ children: ReactNode }> = ({ children }) => {
   )
 }
 
-const SideNavLink: React.FC<ILinkProps> = ({ children, ...props }) => {
+const SideNavLink = ({ children, ...props }) => {
```
Comment (pettinarip) [inline src/components/SideNavMobile.tsx]: not sure about this one but I think this should still work
```suggestion
const SideNavLink = ({ children, ...props }: ILinkProps) => {
```

File: src/layouts/Docs.tsx [hunk 25]
```diff
@@ -1,3 +1,288 @@
...
+            //   editPath={absoluteEditPath}
+            //   langContributors={allCombinedTranslatorsJson.nodes}
+            // />
+          ) : (
+            // TODO: Implement file contributors
+            // <GitHubContributors
+            //   relativePath={relativePath}
+            //   editPath={absoluteEditPath}
+            // />
+            <Text>GitHubContributors</Text>
```
Comment (pettinarip) [inline src/layouts/Docs.tsx]: This component is already implemented in the tutorials, if we update the branch we should be able to implement it here as well. Not a blocker, we can do it in a separate PR.

File: src/layouts/Docs.tsx [hunk 26]
```diff
@@ -1,3 +1,288 @@
...
+    borderColor="border"
+    {...props}
+  >
+    <Link to="#top">
+      <Translation id="back-to-top" /> ↑
+    </Link>
+  </Flex>
+)
+
+export const docsComponents = {
```
Comment (pettinarip) [inline src/layouts/Docs.tsx:187]: We could clean this up a bit since we are already declaring some of these components in the [`MdComponents` file](https://github.com/ethereum/ethereum-org-fork/blob/dev/src/components/MdComponents.tsx). Those are the default components that are included in every md generated page. I'll push a commit.

File: src/components/SectionNav.tsx
Comment (pettinarip) [inline src/components/SectionNav.tsx:1]: Do you know why this component was not in the fork?

I mean, I see [this component in the main repo](https://github.com/ethereum/ethereum-org-website/blob/dev/src/components/SectionNav.tsx) but I don't see it here. It should be. Probably a bad merge in the syncs...not sure.

**Review by pettinarip — COMMENTED** [review]
Nice job @corwintines. Left a couple of code comments.

On the UI side, I see that there is a small different in the size of the sidenav and a different padding or margin the content. Check it out:

Fork version
image

Prod version
image

File: next.config.js (hunk 20 above)
Comment (corwintines) [reply next.config.js]: I can remove this comment, sorry was bad copy paste for yaml importing here. I dont think we would need it for JSON, but needed it to properly import YAM

File: public/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashimoto/index.md (hunk 21 above)
Comment (corwintines) [reply public/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashimoto/index.md]: Will try to sort out whats going on here. I was getting an error of a div within a paragraph, but this isn't the right way to do it. Will revert this change and dig into this more.

File: public/content/translations/de/developers/docs/nodes-and-clients/index.md (hunk 22 above)
Comment (corwintines) [reply public/content/translations/de/developers/docs/nodes-and-clients/index.md]: Correct, these arent in the english folder anymore and were causing build errors

File: src/layouts/Docs.tsx (hunk 25 above)
Comment (corwintines) [reply src/layouts/Docs.tsx]: Brought this in!

File: src/layouts/Docs.tsx (hunk 26 above)
Comment (corwintines) [reply src/layouts/Docs.tsx:187]: Which components here do you see needing to be cleaned up? When migrating it over, there are some styles only in the developer docs when it comes to headings, and paragraph. I imported from this file, and just extended them with the styles. Feel free to push what you want cleaned up though.

File: src/components/SectionNav.tsx
Comment (corwintines) [reply src/components/SectionNav.tsx:1]: Im not sure where this component is really used if im honest. I see it wasnt migrated over, but does it need to be is my question after looking at it too.

**Comment by corwintines** [comment]
@pettinarip I have fixed all the bugs here, and a bug in the activeHash that I found when testing

Im seeing parity in terms of design on my end now

Local implementation:

Screen Shot 2023-11-19 at 12 35 03 PM

Production:
Screen Shot 2023-11-19 at 12 35 11 PM

I think this will be good to go now

File: src/layouts/Docs.tsx (hunk 26 above)
Comment (pettinarip) [reply src/layouts/Docs.tsx:187]: I should have clarified. I've already pushed [a commit](https://github.com/ethereum/ethereum-org-fork/pull/114/commits/09c929d617c7d43ab993f35faf67ecb49bd777be) for this, you can check it out and see what I was referring to.

In short, I basically removed some components from the exported mapping that you are doing here because it was redundant. Most of the common components are already defined in the MdComponents file.

File: src/components/SectionNav.tsx
Comment (pettinarip) [reply src/components/SectionNav.tsx:1]: Just checked and can confirm that is not used in the main repo, we can delete it.

File: src/hooks/useActiveHash.ts [hunk 27]
```diff
@@ -25,7 +25,8 @@ export const useActiveHash = (
     )
 
     itemIds?.forEach((id) => {
-      const element = document.getElementById(id)
+      // Remove # from id. EX: #element-id -> element-id
+      const element = document.getElementById(id.replace("#", ""))
```
Comment (pettinarip) [inline src/hooks/useActiveHash.ts:29]: Nice catch

**Review by pettinarip — APPROVED** [review]
Niceee!!! merging after build passes @corwintines gj
